src/track/beats.cpp

```diff
@@ -53,3 +190,676 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+    BeatList::const_iterator on_beat = m_beats.cend();
+    BeatList::const_iterator previous_beat = m_beats.cend();
+    BeatList::const_iterator next_beat = m_beats.cend();
+    for (; it != m_beats.end(); ++it) {
+        qint32 delta = it->frame_position() - beat.frame_position();
+
+        // We are "on" this beat.
+        if (abs(delta) < kFrameEpsilon) {
+            on_beat = it;
+            break;
+        }
+
+        if (delta < 0) {
+            // If we are not on the beat and delta < 0 then this beat comes
+            // before our current position.
+            previous_beat = it;
+        } else {
+            // If we are past the beat and we aren't on it then this beat comes
+            // after our current position.
+            next_beat = it;
+            // Stop because we have everything we need now.
+            break;
+        }
+    }
+
+    // If we are within epsilon frames of a beat then the immediately next and
+    // previous beats are the beat we are on.
+    if (on_beat != m_beats.end()) {
+        next_beat = on_beat;
+        previous_beat = on_beat;
+    }
+
+    if (n > 0) {
+        for (; next_beat != m_beats.end(); ++next_beat) {
+            if (!next_beat->enabled()) {
+                continue;
+            }
+            if (n == 1) {
+                // Return a sample offset
+                return next_beat->frame_position();
+            }
+            --n;
+        }
+    } else if (n < 0 && previous_beat != m_beats.end()) {
+        for (; true; --previous_beat) {
+            if (previous_beat->enabled()) {
+                if (n == -1) {
+                    // Return a sample offset
+                    return previous_beat->frame_position();
+                }
+                ++n;
+            }
+
+            // Don't step before the start of the list.
+            if (previous_beat == m_beats.begin()) {
+                break;
+            }
+        }
+    }
+    return -1;
+}
+
+std::unique_ptr<BeatIterator> Beats::findBeatsNew(FrameNum startFrame, FrameNum stopFrame) const {
+    QMutexLocker locker(&m_mutex);
+    if (!isValid() || startFrame > stopFrame) {
+        return std::unique_ptr<BeatIterator>();
+    }
+
+    track::io::Beat startBeat, stopBeat;
+    startBeat.set_frame_position(startFrame);
+    stopBeat.set_frame_position(stopFrame);
+
+    BeatList::const_iterator firstBeat =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), startBeat, BeatLessThan);
+
+    BeatList::const_iterator lastBeat =
+            std::upper_bound(m_beats.cbegin(), m_beats.cend(), stopBeat, BeatLessThan);
+    if (lastBeat >= m_beats.cbegin()) {
+        lastBeat = m_beats.cend() - 1;
+    }
+
+    if (firstBeat >= lastBeat) {
+        return std::unique_ptr<BeatIterator>();
+    }
+    return std::make_unique<BeatIterator>(firstBeat, lastBeat);
+}
+
+bool Beats::hasBeatInRangeNew(double startSample, double stopSample) const {
```

Comment:
Some functions expect arguments in frames, some in samples?

Reply:
The original code was using samples.
I was advised to change the logic to use frames. So I did.
I leaved the original function (using samples) as adaptors that call the new ones.
The idea is to have all tests passing (now I have 31 test errors) and when this is OK will modify the code calling the adaptors to use frames instead of samples and calling the new functions.
After that I will remove the adaptors an rename the methods to have the original names.
Makes sense?